### PR TITLE
cc-wrapper: add libcxxabi include flag for LLVM

### LINKED ISSUE #255487: '__cxxabi_config.h' file not found with clang-wrapper
### Describe the bug
Clang's wrapper does not set `${libcxxabi.dev}/c++/v1` as an include path. Compiling files with <cxxabi.h> fails since <cxxabi.h> requires <__cxxabi_config.h> which is present in the aforementioned path

### Minimal Reproduction
```cc
#if __has_include(<cxxabi.h>)
#  include <cxxabi.h>
#endif


int main() {
    return 0;
}
```

### Current ouput
`clang++ alo.cc`
```
In file included from alo.cc:2:
/nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include/cxxabi.h:20:10: fatal error: '__cxxabi_config.h' file not found
#include <__cxxabi_config.h>
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
```
#### Verbose
`clang++ --verbose alo.cc`
```
clang version 16.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /nix/store/09ccpv2ny36nfjigv260b2m8rpxccy2j-clang-16.0.6/bin
Found candidate GCC installation: /nix/store/n8dryz4xf7ln028j37zapgzg4j47p743-gcc-12.3.0/lib/gcc/x86_64-unknown-linux-gnu/12.3.0
Found candidate GCC installation: /nix/store/n8dryz4xf7ln028j37zapgzg4j47p743-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0
Selected GCC installation: /nix/store/n8dryz4xf7ln028j37zapgzg4j47p743-gcc-12.3.0/lib64/gcc/x86_64-unknown-linux-gnu/12.3.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/nix/store/09ccpv2ny36nfjigv260b2m8rpxccy2j-clang-16.0.6/bin/clang-16" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name alo.cc -mrelocation-model pic -pic-level 2 -fhalf-no-semantic-interposition -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -v -fcoverage-compilation-dir=/home/natto -nostdsysteminc -resource-dir /nix/store/yfw10j91pd0c4aycascgfirq58nwik03-clang-wrapper-16.0.6/resource-root -idirafter /nix/store/4lap45ifj50lp980id60fnbvs81yj9lr-glibc-2.37-8-dev/include -isystem /nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include -isystem /nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include -isystem /nix/store/fqvx4m5d12jpickwd1jm3lmqf55xq4ln-compiler-rt-libc-16.0.6-dev/include -isystem /nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include -isystem /nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include -isystem /nix/store/fqvx4m5d12jpickwd1jm3lmqf55xq4ln-compiler-rt-libc-16.0.6-dev/include -isystem /nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include/c++/v1 -U _FORTIFY_SOURCE -U _FORTIFY_SOURCE -D _FORTIFY_SOURCE=2 -internal-isystem /nix/store/yfw10j91pd0c4aycascgfirq58nwik03-clang-wrapper-16.0.6/resource-root/include -source-date-epoch 315532800 -O2 -Wformat -Wformat-security -Werror=format-security -fdeprecated-macro -fdebug-compilation-dir=/home/natto -ferror-limit 19 -fwrapv -stack-protector 2 -stack-protector-buffer-size 4 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/nix-shell.kawEoa/nix-shell.3Bf8zw/alo-4bb3ab.o -x c++ alo.cc
clang -cc1 version 16.0.6 based upon LLVM 16.0.6 default target x86_64-unknown-linux-gnu
ignoring duplicate directory "/nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include"
ignoring duplicate directory "/nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include"
ignoring duplicate directory "/nix/store/fqvx4m5d12jpickwd1jm3lmqf55xq4ln-compiler-rt-libc-16.0.6-dev/include"
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include
 /nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include
 /nix/store/fqvx4m5d12jpickwd1jm3lmqf55xq4ln-compiler-rt-libc-16.0.6-dev/include
 /nix/store/2m1m4lwcr89xwyli8ir6bzkp23ymz738-libcxx-16.0.6-dev/include/c++/v1
 /nix/store/yfw10j91pd0c4aycascgfirq58nwik03-clang-wrapper-16.0.6/resource-root/include
 /nix/store/4lap45ifj50lp980id60fnbvs81yj9lr-glibc-2.37-8-dev/include
End of search list.
In file included from alo.cc:2:
/nix/store/wk6mn10acxsi3q3hxzhsypwxlf5nmkxa-libcxxabi-16.0.6-dev/include/cxxabi.h:20:10: fatal error: '__cxxabi_config.h' file not found
#include <__cxxabi_config.h>
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
```

### Notes
I created the environment using this minimal mkShell expression
```nix
mkShell.override { stdenv = llvmPackages_16.libcxxStdenv; } { }
```

### Proposed fix
Adding `-isystem ${lib.getDev libcxx.cxxabi}/include/c++/v1` to `libcxx-cxxflags`

### PR DESCRIPTION
**Pull request by natto1784**
## Description of changes
Should fix https://github.com/NixOS/nixpkgs/issues/255487

## Things done
Added `-isystem ${libcxxabi}/include/c++/v1` to `libcxx-cxxflags` for clang

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/test/cc-wrapper/default.nix [hunk 1]
```diff
@@ -30,6 +30,13 @@ in stdenv.mkDerivation {
     $CXX -o cxx-check ${./cxx-main.cc}
     ${emulator} ./cxx-check
 
+    # test for https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1431745905
+    # .../include/cxxabi.h:20:10: fatal error: '__cxxabi_config.h' file not found
+    # in libcxxStdenv
+    echo "checking whether cxxabi.h can be included... " >&2
+    $CXX -o include-cxxabi ${./include-cxxabi.cc}
```
Comment (Artturin) [inline pkgs/test/cc-wrapper/default.nix:37]: This pr fixes this test for `tests.cc-wrapper.llvmTests.llvmPackages_{15,16}.libcxx`

File: pkgs/development/compilers/llvm/16/default.nix [hunk 2]
```diff
@@ -255,14 +255,6 @@ in let
...
           "-Wno-unused-command-line-argument"
           "-B${targetLlvmLibraries.compiler-rt}/lib"
-
-          # Combat "__cxxabi_config.h not found". Maybe this could be fixed by
-          # copying these headers into libcxx? Note that building libcxx
-          # outside of monorepo isn't supported anymore, might be related to
-          # https://github.com/llvm/llvm-project/issues/55632
-          # ("16.0.3 libcxx, libcxxabi: circular build dependencies")
-          # Looks like the machinery changed in https://reviews.llvm.org/D120727.
-          "-I${lib.getDev targetLlvmLibraries.libcxx.cxxabi}/include/c++/v1"
```
Comment (Artturin) [inline pkgs/development/compilers/llvm/16/default.nix:265]: removed old workaround

**Comment by Artturin** [comment]
I have to consider whether we should do this or

`install -m 644 ../../${pname}/include/__cxxabi_config.h "$dev/include"` in `libcxxabi` `postInstall`, were already copying `cxxabi.h` in there

File: pkgs/build-support/cc-wrapper/default.nix [hunk 3]
```diff
@@ -469,6 +469,7 @@ stdenv.mkDerivation {
     ''
     + optionalString (libcxx.isLLVM or false) ''
       echo "-isystem ${lib.getDev libcxx}/include/c++/v1" >> $out/nix-support/libcxx-cxxflags
```
Comment (Artturin) [inline pkgs/build-support/cc-wrapper/default.nix:471]: Before llvm 15 `__cxxabi_config.h` was included in `libcxx` so this worked correctly

```
llvmPackages_13.libcxx.dev                        3,133 r /nix/store/5wfncnq3svyf54hfahi44smavyfbhq5b-libcxx-13.0.1-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_7.libcxx.dev                         2,278 r /nix/store/2i745a6pvd2qbi0lf9a61ia0svvg9q0b-libcxx-7.1.0-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_14.libcxx.dev                        3,142 r /nix/store/jri1wy9fxnnln4wv7dmzdhs9w5ll4jfa-libcxx-14.0.6-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_15.libcxxabi.dev                     3,266 r /nix/store/bbvhnvq66lr6yzqxqbzff2341h9xg332-libcxxabi-15.0.7-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_6.libcxx.dev                         2,032 r /nix/store/bpyj2sk6mh9vlli539jsr0w45rpig5pc-libcxx-6.0.1-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_5.libcxx.dev                         2,032 r /nix/store/fciabccmn17ijmfxxn50diin0dcjz7m6-libcxx-5.0.2-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_8.libcxx.dev                         2,278 r /nix/store/8sg9d3j38i3p13sg7pp4bc44szqakj7s-libcxx-8.0.1-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_9.libcxx.dev                         2,411 r /nix/store/1b5yn18w09n9r50i9162zvn3zq3jn98k-libcxx-9.0.1-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_16.libcxxabi.dev                     3,266 r /nix/store/02wpjmp2zjjxz13z7g599mniwi25zkcy-libcxxabi-16.0.6-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_10.libcxx.dev                        2,500 r /nix/store/j41yp16y29b7gw2bd5c1ljfyyykgsh12-libcxx-10.0.1-dev/include/c++/v1/__cxxabi_config.h
llvmPackages_12.libcxx.dev                        3,133 r /nix/store/fyr084bq8496099nllfjzapis7mxpj39-libcxx-12.0.1-dev/include/c++/v1/__cxxabi_config.h
```

and
cxxabi.h was in both libcxx and libcxxabi
```
llvmPackages_13.libcxx.dev                        7,215 r /nix/store/5wfncnq3svyf54hfahi44smavyfbhq5b-libcxx-13.0.1-dev/include/c++/v1/cxxabi.h
llvmPackages_7.libcxx.dev                         6,945 r /nix/store/2i745a6pvd2qbi0lf9a61ia0svvg9q0b-libcxx-7.1.0-dev/include/c++/v1/cxxabi.h
llvmPackages_14.libcxx.dev                        7,215 r /nix/store/jri1wy9fxnnln4wv7dmzdhs9w5ll4jfa-libcxx-14.0.6-dev/include/c++/v1/cxxabi.h
libcxxrt.out                                      9,040 r /nix/store/sdw3q6nzxgdh6n6jdnnsgm7jpg5wsp1v-libcxxrt-unstable-2022-08-08/include/cxxabi.h
llvmPackages_12.libcxxabi.dev                     7,215 r /nix/store/6lzf327lxdi743i8n8s3igq962jcp5r7-libcxxabi-12.0.1-dev/include/cxxabi.h
llvmPackages_15.libcxxabi.dev                     7,237 r /nix/store/bbvhnvq66lr6yzqxqbzff2341h9xg332-libcxxabi-15.0.7-dev/include/c++/v1/cxxabi.h
llvmPackages_15.libcxxabi.dev                     7,237 r /nix/store/bbvhnvq66lr6yzqxqbzff2341h9xg332-libcxxabi-15.0.7-dev/include/cxxabi.h
llvmPackages_5.libcxxabi.dev                      6,945 r /nix/store/qlzdwhhd0x9fda4mk1a83jmmxza4a64p-libcxxabi-5.0.2-dev/include/cxxabi.h
libcxxabi.dev                                     7,029 r /nix/store/h84lr9b4q34f37jvqzqkkf4gnrlrgahl-libcxxabi-11.1.0-dev/include/cxxabi.h
llvmPackages_14.libcxxabi.dev                     7,215 r /nix/store/zkfgj0hy0za0yhwnrfdndvzr301aj4pq-libcxxabi-14.0.6-dev/include/cxxabi.h
llvmPackages_6.libcxx.dev                         6,945 r /nix/store/bpyj2sk6mh9vlli539jsr0w45rpig5pc-libcxx-6.0.1-dev/include/c++/v1/cxxabi.h
llvmPackages_5.libcxx.dev                         6,945 r /nix/store/fciabccmn17ijmfxxn50diin0dcjz7m6-libcxx-5.0.2-dev/include/c++/v1/cxxabi.h
llvmPackages_8.libcxx.dev                         6,987 r /nix/store/8sg9d3j38i3p13sg7pp4bc44szqakj7s-libcxx-8.0.1-dev/include/c++/v1/cxxabi.h
llvmPackages_9.libcxx.dev                         7,023 r /nix/store/1b5yn18w09n9r50i9162zvn3zq3jn98k-libcxx-9.0.1-dev/include/c++/v1/cxxabi.h
llvmPackages_16.libcxxabi.dev                     7,237 r /nix/store/02wpjmp2zjjxz13z7g599mniwi25zkcy-libcxxabi-16.0.6-dev/include/c++/v1/cxxabi.h
llvmPackages_16.libcxxabi.dev                     7,237 r /nix/store/02wpjmp2zjjxz13z7g599mniwi25zkcy-libcxxabi-16.0.6-dev/include/cxxabi.h
llvmPackages_10.libcxxabi.dev                     7,029 r /nix/store/pwx93kc21scj2ikif98d9zzyzaicag8x-libcxxabi-10.0.1-dev/include/cxxabi.h
llvmPackages_7.libcxxabi.dev                      6,945 r /nix/store/wx2xdg98ddhrs5iif85w8dgmskgkmjv2-libcxxabi-7.1.0-dev/include/cxxabi.h
llvmPackages_9.libcxxabi.dev                      7,023 r /nix/store/lhjbqxfqqckm3x2ci9v7vgblmw6jr2jc-libcxxabi-9.0.1-dev/include/cxxabi.h
llvmPackages_6.libcxxabi.dev                      6,945 r /nix/store/lc41k1nfaslcp8gbgl948ksx52n19j20-libcxxabi-6.0.1-dev/include/cxxabi.h
llvmPackages_8.libcxxabi.dev                      6,987 r /nix/store/khf9r8s7llkmpsxza19pqavr26j3w6lb-libcxxabi-8.0.1-dev/include/cxxabi.h
llvmPackages_10.libcxx.dev                        7,029 r /nix/store/j41yp16y29b7gw2bd5c1ljfyyykgsh12-libcxx-10.0.1-dev/include/c++/v1/cxxabi.h
llvmPackages_12.libcxx.dev                        7,215 r /nix/store/fyr084bq8496099nllfjzapis7mxpj39-libcxx-12.0.1-dev/include/c++/v1/cxxabi.h
```

**Comment by Artturin** [comment]
> I have to consider whether we should do this or
> 
> `install -m 644 ../../${pname}/include/__cxxabi_config.h "$dev/include"` in `libcxxabi` `postInstall`, were already copying `cxxabi.h` in there

basically if we do this then we don't have to special case cxxabi in the wrapper and can benefit from the hook https://github.com/NixOS/nixpkgs/blob/fdebb755af09943202242371f0e5a79b6c7332c5/pkgs/build-support/cc-wrapper/setup-hook.sh#L71

**Comment by Artturin** [comment]
Let's do this

TODO remove now should be unnecessary postInstalls from all libcxxabi's

**Comment by Artturin** [comment]
@natto1784 do you want to resign the first commit before merge.

CC @pwaller

**Review by Artturin — APPROVED** [review]
This restores llvm <15 behaviour

**Comment by natto1784** [comment]
> Let's do this
> 
> TODO remove now should be unnecessary postInstalls from all libcxxabi's

Assuming this is redundant
```nix
 install -m 644 ../../${pname}/include/${if stdenv.isDarwin then "*" else "cxxabi.h"} "$dev/include"
```
I am curious as to why is everything installed from the $src itself for darwin but not other platforms



> @natto1784 do you want to re-sign the first commit before merge.
> 
> CC @pwaller

Oh yeah sure

**Comment by Artturin** [comment]
> > Let's do this
> > TODO remove now should be unnecessary postInstalls from all libcxxabi's
> 
> Assuming this is redundant
> 
> ```nix
>  install -m 644 ../../${pname}/include/${if stdenv.isDarwin then "*" else "cxxabi.h"} "$dev/include"
> ```
> 
> I am curious as to why is everything installed from the $src itself for darwin but not other platforms

I'm going to leave that for later because removing it is not necessary in this PR.

**Pull request merged**